### PR TITLE
Add support for ARM64 on LGBM

### PR DESCRIPTION
**Pull request by artpdr**
Make it possible to use LGBM on nodes with ARM64 architectures.

### CODE REVIEW

**Comment by artpdr** [comment]
Hey guys. I still some comments in the files that need to be addressed prior to merge... However, can anyone create a xp tag from this PR and push it so that I can test the artifacts?

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -24,13 +24,18 @@ jobs:
...
+        uses: actions/setup-java@v3
         with:
-          java-version: 1.8
+          java-version: '8'
+          distribution: 'zulu'
           server-id: ossrh
           server-username: MAVEN_USERNAME
           server-password: MAVEN_PASSWORD
 
+      - uses: docker/setup-qemu-action@v2
```
Comment (AlbertoEAF) [inline .github/workflows/deploy.yml:35]: do we need this for deploy? isn't the jar built by then?

**Review by AlbertoEAF — APPROVED** [review]
LGTM and the UATs have already been conducted in a real cluster, with a large dataset, and all the score errors are still < 1E-14 between ARM64 and AMD64, so :+1:

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * © 2023 Feedzai, Strictly Confidential
+ */
+package com.feedzai.openml.provider.lightgbm;
+
+/**
+ * Enum that represents the cpu architecture where code is running and consequent lgbm native libs locations.
+ */
+public enum CpuArchitecture {
+    AARCH64("arm64"),
+    AMD64("amd64");
```
Comment (gandola) [inline openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java:16]: I'm a bit worried about this: https://github.com/openjdk/jdk/blob/master/make/autoconf/platform.m4#L481C2-L481C2

it maybe not be critical it now use case but I'm afraid we are not capturing all the possibilities on this and then we need to deal with it getting all the context on this. At least some documentation why are we only considering these 2 options.

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java [hunk 3]
```diff
@@ -57,9 +57,11 @@ static public void loadLibs() {
 
         if (!libsLoaded) {
             try {
-                loadSharedLibraryFromJar("libgomp.so.1.0.0");
-                loadSharedLibraryFromJar("lib_lightgbm.so");
-                loadSharedLibraryFromJar("lib_lightgbm_swig.so");
+                final String cpuArchName = System.getProperty("os.arch");
+                final CpuArchitecture cpuArchitecture = CpuArchitecture.valueOf(cpuArchName.toUpperCase());
```
Comment (gandola) [inline openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:61]: If a non-supported arch is used an exception will be thrown, however, I would catch it and add some meaningful message to it to help the users and then re-throw it.

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java [hunk 4]
```diff
@@ -57,9 +57,11 @@ static public void loadLibs() {
...
+                loadSharedLibraryFromJar("libgomp.so.1.0.0", cpuArchitecture);
+                loadSharedLibraryFromJar("lib_lightgbm.so", cpuArchitecture);
+                loadSharedLibraryFromJar("lib_lightgbm_swig.so", cpuArchitecture);
```
Comment (gandola) [inline openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:64]: Why are we including SO libraries within this jar? why aren't they separated artifacts?

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java [hunk 5]
```diff
@@ -57,9 +57,11 @@ static public void loadLibs() {
...
             } catch (final IOException e) {
```
Comment (gandola) [inline openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:65]: `ex`

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -30,20 +30,29 @@ jobs:
...
           restore-keys: ${{ runner.os }}-m2
 
+      - uses: docker/setup-qemu-action@v2
+        with:
+          platforms: "arm64"
+
       - name: Build and test
         run: mvn test -B
 
+      # The skipped tests is because h2o-xgboost isn't ready to use on arm64
```
Comment (gandola) [inline .github/workflows/build.yml:52]: Is there any assertion somewhere that prevent users to use `h2o-xgboost` with `arm64`.

**Review by gandola — COMMENTED** [review]
I feel that we need some documentation (README) about this, limitations, etc...

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java (hunk 2 above)
Comment (artpdr) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java:16]: We just support lightgbm compiled for amd64 and aarch64. From the tests I've done on the two architectures, I saw that the JVM set the os.arch to either amd64 on x86_64 or aarch64 on aarch64 cpus... I've always used zulu on the tests with arm64 and openjdk on x86_64.

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java (hunk 3 above)
Comment (artpdr) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:61]: done

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java (hunk 5 above)
Comment (artpdr) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:65]: done

File: .github/workflows/build.yml (hunk 6 above)
Comment (artpdr) [reply .github/workflows/build.yml:52]: no, there isn't

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java (hunk 2 above)
Comment (gandola) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java:16]: Please make sure that is  very well captured on the documentation.

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java (hunk 4 above)
Comment (gandola) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:66]: Well, if that is the approach... but IMO this is not correct. It will increase the jar size significantly and it is not taking advantage of the shared libraries of Operating Systems, this looks like statically linking these libraries. 

Another thing is that each time something is fixed on those libraries this jar will need to be created, etc...

Anyway, if this is the way that it is deployed, I leave that decision to the team.

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java [hunk 7]
```diff
@@ -57,29 +60,51 @@ static public void loadLibs() {
...
             }
 
             logger.info("Loaded LightGBM libs.");
             libsLoaded = true;
         }
     }
 
+    static CpuArchitecture getCpuArchitecture(final String cpuArchName) {
+        final String upperCaseCpuArchName = cpuArchName.toUpperCase();
+        final boolean knownCpuArchitecture = Arrays.stream(CpuArchitecture.values())
```
Comment (gandola) [inline openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:78]: We can simply catch the `IllegalArgumentException` thrown by the  `CpuArchitecture.valueOf(upperCaseCpuArchName)` and work from there no? It is simpler I guess.

Comment (artpdr) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:78]: it is. done

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java (hunk 4 above)
Comment (artpdr) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:66]: Again, it's how it was.  We are more concerned in making lgbm working on amd64 than in changing the structure how this is done. @AlbertoEAF can you give more context why this is done this way? Also @gandola the only thing I see that may be removed from here is the libgomp which we could use from the OS... right?

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java (hunk 2 above)
Comment (artpdr) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/CpuArchitecture.java:16]: It's in the README

**Review by gandola — APPROVED** [review]
:+1:

File: openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java (hunk 4 above)
Comment (AlbertoEAF) [reply openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java:66]: Hi there :)

This is the best approach possible given our goals: low total cost of ownership, ease of deployment and reliability.  We put the dynamic libraries inside the (fat-)jar **on purpose** so it is self-sufficient in any deployment scenario.

This started for historical reasons and by design so we did not depending at all on the environment/setup of the target cluster/pulse machines used when we were mostly on-prem and clients managed those installations -- it would not be feasible at all to assume clients would properly manage the available shared libs on their cluster machines,  etc.; and if they did, the time for those changes to happen would be in the order of months, and the reliability of such setups questionable and prone to generating issues for our teams.

Now with the move to cloud this mechanism again proved simple and reliable. We didn't require coordination with other teams, and it always works. Additionally, with this strategy and from this MR onward we will be able to run on our cloud with both AMD64 and ARM64 machines without any configuration required, nor Pulse knowing at all whether the target EMR machines are AMD64 or ARM64.

We need to keep libgomp as well, as several of these machines don't have it available. This has been extensively tested on various clusters and without this library available at all times we'd be forced to fall back to a single thread for training (again, not viable).

The only requirement of this approach (so the .so's we ship are "universal") is that our compiled shared libraries be compatible with the C runtime of the target machines (glibc target machine >= glibc using during compile time of these .so's). That is why when we compile them we use an old docker environment that is older than any relevant linux distro in use.


Should we document this on the project @gandola ? Maybe add a summary of this to the docstring?
